feature_engine/dataframe_checks.py

```diff
@@ -129,3 +130,77 @@ def _check_contains_inf(X: pd.DataFrame, variables: List[Union[str, int]]) -> No
             "Some of the variables to transform contain inf values. Check and "
             "remove those before using this transformer."
         )
+
+
+def _check_pd_X_y(
```

Comment:
Hi guys, I spent a good few hours going over this function, and I still can't decide what is best.

The thing is, if we are going to replace `is_dataframe` with this function, then the function should be able to return a copy of the df (like `is_dataframe`) and potentially a copy of y (not sure is necessary), to avoid inadvertently modifying the user's data. So before merging, at least, we need to ensure we return a copy of X.

When playing around with this  function, I found out some errors in the tests raised by check_estimator. One of the tests would be if y=None, and I've noticed that we do not handle this situation. At the moment, we are assuming that the user would enter a numpy array or a series, but in theory, they could also enter a list, a tuple or None, and all of that would work just fine with this version of check_pd_x_y and still fail in the concatenation, unless we ensure we return a pd.series from this function.

But, pd.Series(None) would still return a pd.series. And if we do pd.Series(None, index=X.index) it will return a pd series of the length of the dataframe full of nan, when the user enters None.

If I convert a numpy array to a df, and force the index of the dataframe, what if the array was shorter than the df? would it fail or would it introduce Nan? I did not check this, so thinking out loud.

I then went ahead and checked the check_x_y function from sklearn for some inspiration, and I see that they handle the y=None at the very top of the check. So we probably should do that as well.

Reply:
> If I convert a numpy array to a df, and force the index of the dataframe, what if the array was shorter than the df? would it fail or would it introduce Nan? I did not check this, so thinking out loud.

Pretty sure I check for incompatible dimensions and unit test it, but I'll double check.

I'll go ahead and make the following changes:
- Return copies of X and y when they are incoming as pandas objects
- Raise an exception if y (or X) is None
- Accommodate other array-like inputs for y

Sound good?